File: lecture_21/p7.cpp
#include<iostream>
#include<vector>
#include<queue>
using namespace std;

int longestOnes(vector<int>& arr, int k) {
    int ans=0;
    queue<int> idx_que;
    int j=0;
    for(int i=0;i<arr.size();i++)
    {
        if(arr[i]==0)
        {
            idx_que.push(i);
            if(idx_que.size()>k){
                j=idx_que.front()+1; 
                idx_que.pop();
            }
        }
        ans=max(ans,i-j+1);
    }
    return ans;
}
int main(int args,char** argv)
{
    vector<int> arr={0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1};
    int k=3;
    cout<<longestOnes(arr,k);
    return 0;
}

File: lecture_24/p7.java
import java.util.Stack;

public class p7 {
    public static int largestRectangleArea(int[] arr) {
        Stack<Integer> st=new Stack<>();
        st.push(-1);
        int ans=0;
        for(int i=0;i<arr.length;i++)
        {
            while(st.peek()!=-1 && arr[i]<arr[st.peek()])
            {
                int a=arr[st.pop()] * (i-st.peek()-1);
                ans=Math.max(ans,a);
            }
            st.push(i);
        }
        while(st.peek()!=-1)
        {
            int a=arr[st.pop()] * (arr.length-st.peek()-1);
            ans=Math.max(ans,a);
        }
        return ans;
    }
    public static int maximalRectangle(char[][] matrix) {
        if(matrix.length==0)
            return 0;
        int ans=0;
        int arr[]=new int[matrix[0].length];
        for(int i=0;i<matrix.length;i++)
        {
            for(int j=0;j<matrix[0].length;j++){
                if(matrix[i][j]=='0')
                    arr[j]=0;
                else
                    arr[j]+=1;
            }
            ans=Math.max(ans, largestRectangleArea(arr));
        }
        return ans;
    }
    public static void main(String[] args) {
        char matrix[][]={
            {'1','0','1','0','0'},
            {'1','0','1','1','1'},
            {'1','1','1','1','1'},
            {'1','0','0','1','0'}
        };
        System.out.println(maximalRectangle(matrix));
    }
}